src/de.rs

```diff
@@ -305,59 +275,69 @@ impl<'de, 'a> de::Deserializer<'de> for &'a mut Deserializer<'de> {
     fn deserialize_option<V>(self, visitor: V) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        if self.input.starts_with("null") {
-            self.input = &self.input["null".len()..];
-            visitor.visit_none()
-        } else {
-            visitor.visit_some(self)
+        match seq(b"null").discard().parse(&mut self.input) {
+            Ok(_) => visitor.visit_none(),
+            Err(_) => visitor.visit_some(self),
         }
     }
 
     // In Serde, unit means an anonymous value containing no data.
     fn deserialize_unit<V>(self, visitor: V) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        if self.consume("()") {
-            visitor.visit_unit()
-        } else {
-            Err(Error::ExpectetUnit)
+        match self.consume("()") {
+            Ok(_) => visitor.visit_unit(),
+            Err(_) => Err(Error::ExpectetUnit),
         }
     }
 
     fn deserialize_unit_struct<V>(
         self,
-        _name: &'static str,
+        name: &'static str,
         visitor: V
     ) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        // TODO: Should we also allow `MyUnitValue`?
-
-        self.deserialize_unit(visitor)
+        match self.consume(name) {
+            Ok(_) => visitor.visit_unit(),
+            Err(_) => Err(Error::ExpectedStructName),
+        }
     }
 
     fn deserialize_newtype_struct<V>(
         self,
-        _name: &'static str,
+        name: &'static str,
         visitor: V
     ) -> Result<V::Value>
         where V: Visitor<'de>
     {
-        visitor.visit_newtype_struct(self)
+        match self.consume(name) {
+            Ok(_) => match self.consume("(") {
+                Ok(_) => {
+                    let value = visitor.visit_newtype_struct(&mut *self)?;
+                    let _ = comma().parse(&mut self.input);
+                    self.consume(")")
+                        .map(|_| value)
+                        .map_err(|_| Error::ExpectedStructEnd)
+                },
+                Err(_) => Err(Error::ExpectedStruct),
+            },
+            Err(_) => visitor.visit_newtype_struct(self),
```

Comment:
I'd say screw it. `9.0` is inconsistent with the design, so I'd rather just support omitting the struct name as a general rule and be done with it. Hence, no `0.9`, but `(0.9)` is fine. That's what the latest code does.